controller/src/lib.rs

```diff
@@ -20,114 +21,184 @@ pub mod plan;
+                );
+
+            Ok((
+                ScheduleResponse::Scheduled {
+                    drone: drone.clone(),
+                    backend_id: spawn_request.backend_id,
+                    bearer_token: spawn_request.bearer_token.clone(),
+                    spawned: true,
+                },
+                Some(seq_id),
+            ))
+        }
+        Ok(false) => {
+            tracing::warn!("Drone rejected backend.");
+            Ok((ScheduleResponse::NoDroneAvailable, None))
+        }
+        Err(error) => {
+            tracing::warn!(?error, "Scheduler returned error.");
+            Ok((ScheduleResponse::NoDroneAvailable, None))
+        }
+    }
+}
+
+/// get backend associated with a lock or error
+fn backend_of_lock(
+    state: &StateHandle,
+    cluster_name: &ClusterName,
+    lock: &str,
+) -> anyhow::Result<BackendId> {
+    state
+        .state()
+        .cluster(cluster_name)
+        .ok_or_else(|| anyhow!("no cluster"))?
+        .locked(lock)
+        .ok_or_else(|| anyhow!("no backend"))
+}
+
+fn schedule_response_for_existing_backend(
+    state: &StateHandle,
+    cluster: ClusterName,
+    backend: BackendId,
+) -> anyhow::Result<ScheduleResponse> {
+    // Anything that fails to find the drone results in an error here, since we just
+    // checked that the lock is held which implies that the drone exists.
+    let state = state.state();
+    tracing::info!("getting cluster from state");
+    let cluster = state
+        .cluster(&cluster)
+        .ok_or_else(|| anyhow!("no such cluster"))?;
+    tracing::info!("fetching backend!");
+    let (drone, bearer_token) = {
+        let backend_state = cluster
+            .backend(&backend)
+            .ok_or_else(|| anyhow!("Lock held by a backend that doesn't exist."))?;
+
+        let drone = backend_state
+            .drone
+            .clone()
+            .ok_or_else(|| anyhow!("Lock held by a backend without a drone assignment."))?;
+
+        let bearer_token = backend_state.bearer_token.clone();
+
+        (drone, bearer_token)
+    };
+
+    Ok(ScheduleResponse::Scheduled {
+        drone,
+        backend_id: backend,
+        bearer_token,
+        spawned: false,
+    })
+}
+
+async fn process_response(
+    state: &StateHandle,
+    sr: &ScheduleRequest,
+    scheduler: &Scheduler,
+    nats: &TypedNats,
+) -> anyhow::Result<ScheduleResponse> {
+    tracing::info!("checking locks");
+    let cluster_name = sr.cluster.clone();
+    if let Some(lock_name) = sr.lock.clone() {
+        tracing::info!(?lock_name, "scheduling lock");
+
+        if let Ok(backend) = backend_of_lock(state, &cluster_name, &lock_name) {
+            tracing::info!(?backend, "fetch preexisting backend");
+            return schedule_response_for_existing_backend(state, cluster_name, backend);
+        }
+
+        tracing::info!("spawn with lock");
+        let drone = match scheduler.schedule(&cluster_name, Utc::now()) {
```

Comment:
I'm struggling to see a difference in behavior between the code path here and the code path below here and the `else`-branch, is there one?

If not, we could probably remove the `else` branch and have the code be in the main function body, and have the execution just drop through here in the else-case.

Reply:
Yea all the finicky locking stuff ends up here, It's in the big pr, I'll have the specific one for that up tom.

Reply:
yeah, it's kind of hard to review this without the rest of the context :/

Reply:
hmm, I think I'll minimize this one to just do the spawn per schedule req (I figured that was mostly orthogonal so I could separate it out)